### ISSUE TITLE
Uncaught Error: Cannot find module './locale'

### ISSUE THREAD

**Issue by mau21mau** [moment/moment#5484]
**Describe the bug**
I get this error on chrome console after I build the production optimized build

Uncaught Error: Cannot find module './locale'
    at 2.d0944550.chunk.js:1
    at Module.<anonymous> (2.d0944550.chunk.js:1)
    at f ((index):1)
    at Module.837 (main.091f08fc.chunk.js:1)
    at f ((index):1)
    at Object.540 (main.091f08fc.chunk.js:1)
    at f ((index):1)
    at a ((index):1)
    at Array.e [as push] ((index):1)
    at main.091f08fc.chunk.js:1

This is my package.json

```
{
  "name": "sheetgo-front-end",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@allpro/react-router-pause": "^1.1.3",
    "@material-ui/core": "^4.4.2",
    "@material-ui/icons": "^4.5.1",
    "@material-ui/lab": "^4.0.0-alpha.26",
    "@material-ui/styles": "^4.4.1",
    "@material/icon-button": "^3.1.0",
    "@material/react-button": "^0.15.0",
    "@material/react-card": "^0.15.0",
    "@material/react-icon-button": "^0.15.0",
    "@material/react-material-icon": "^0.15.0",
    "@material/react-select": "^0.15.0",
    "@material/react-tab": "^0.15.0",
    "@mdi/js": "^3.4.93",
    "@mdi/react": "^1.1.0",
    "@stripe/react-stripe-js": "^1.1.0",
    "@stripe/stripe-js": "^1.2.0",
    "@svgr/webpack": "^4.1.0",
    "apexcharts": "^3.8.6",
    "axios": "^0.19.2",
    "browser-info": "^1.2.0",
    "chart.js": "^2.7.2",
    "classnames": "^2.2.6",
    "color-sort": "^0.0.1",
    "connected-react-router": "^6.5.2",
    "countup.js": "^1.9.3",
    "cra-append-sw": "^2.7.0",
    "filepond": "4.1.0",
    "filepond-plugin-file-validate-size": "2.1.1",
    "filepond-plugin-file-validate-type": "1.2.2",
    "form-serialize": "^0.7.2",
    "fuse.js": "^3.4.2",
    "history": "^4.10.1",
    "inputmask": "^4.0.0",
    "is-mobile": "^2.0.0",
    "jquery": "^3.3.1",
    "jquery-ui-sortable-npm": "^1.0.0",
    "js-cookie": "^2.2.0",
    "lodash": "^4.17.10",
    "markdown-to-jsx": "^6.10.3",
    "material-components-web": "^3.1.0",
    "material-ui-chip-input": "^1.0.0",
    "md5": "^2.2.1",
    "memoize-one": "^5.0.4",
    "moment": "^2.22.2",
    "moment-timezone": "^0.5.27",
    "onecolor": "^3.1.0",
    "p-queue": "^6.1.1",
    "payform": "^1.2.2",
    "promise-throttle": "^1.0.0",
    "prop-types": "^15.7.0",
    "query-string": "^6.5.0",
    "react": "^16.8.1",
    "react-apexcharts": "^1.3.3",
    "react-dom": "^16.8.1",
    "react-draggable": "4.2.0",
    "react-filepond": "7.0.1",
    "react-google-authorize": "^1.0.4",
    "react-hotjar": "^2.0.0",
    "react-id-swiper": "^1.6.8",
    "react-intl": "^2.5.0",
    "react-modal-video": "^1.2.3",
    "react-redux": "^7.1.1",
    "react-resizable": "^1.8.0",
    "react-router": "^5.1.2",
    "react-router-dom": "^5.1.2",
    "react-router-last-location": "^2.0.1",
    "react-scripts": "^2.1.4",
    "react-sizeme": "^2.5.2",
    "react-stripe-elements": "^4.0.0",
    "react-text-mask": "^5.4.3",
    "react-virtualized-auto-sizer": "^1.0.2",
    "react-vis-network": "^1.0.0",
    "react-window": "^1.8.1",
    "recompose": "^0.30.0",
    "redux": "^4.0.4",
    "redux-react-session": "^2.4.0",
    "redux-saga": "^1.1.1",
    "redux-thunk": "^2.3.0",
    "socket.io-client": "^2.3.0",
    "sort-by": "^1.2.0",
    "tether": "^1.4.4",
    "throttle-debounce": "^2.1.0",
    "url-loader": "^1.0.1"
  }
```

If I build locally with yarn for the dev version it works

**To Reproduce**
Build an app with those requirements. Should get that error on console.

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Moment-specific environment**

* The time zone setting of the machine the code is running on
* The time and date at which the code was run
* Other libraries in use (TypeScript, Immutable.js, etc)

Please run the following code in your environment and include the output:
```
console.log((new Date()).toString())
console.log((new Date()).toLocaleString())
console.log((new Date()).getTimezoneOffset())
console.log(navigator.userAgent)
console.log(moment.version)
```

**Additional context**

**Comment by manosriram** [comment]
Exactly. I just faced this error and saw that they just pushed a new version hours ago.

**Comment by rilyu** [comment]
if you use yarn, add  resolutions block to package.json to override versions of sub-dependency, it work for me
```
  "resolutions": {
    "moment": "2.24.0"
  }
```

**Comment by aeroxy** [comment]
Same here!

**Comment by aeroxy** [comment]
Also 2.24.0 still has the warning: `./node_modules/chart.js/node_modules/moment/src/lib/locale/locales.js
Critical dependency: the request of a dependency is an expression`

**Comment by brianshen1990** [comment]
Just reproduce this problem

**Comment by nguyenletienbao99** [comment]
Same too!

**Comment by socionity** [comment]
Same here!

**Comment by veryStarters** [comment]
+ 1

**Comment by saadzr** [comment]
Same issue, 

Fixed by removing carret sign before version number. Package.json entry now looks like:
   "moment": "2.24.0"

It looks like a bug from the new version.

**Comment by wktdev** [comment]
Tonight I installed Jest and it created a problem and I was instructed to remove any yarn.lock and package.lock files along with my modules folder and do yarn install or npm install. When I did yarn install I received the error mentioned in this thread and the error mentioned moment-timezone. When I removed my lock files , module folder and reinstalled using npm install  -   everything worked.

**Comment by kothari775** [comment]
Please remove ^ in package.json this way its worked for me 👍 

strange such incident happened early this week due to is-promise library there also issue like ^ need to remove. I believe people will now more or less do for all package.json strict version instead of ^

**Comment by purefan** [comment]
Tried removing the caret from package.json and running + `rm -rf node_modules` + `yarn` and I still get this problem when running webpack

**Comment by ajayar** [comment]
I am using moment-timezone, and it defaults to 2.25. removing caret makes no difference

**Comment by JackLian** [comment]
Same here!

**Comment by constlhq** [comment]
This works for me, I am using create-react-app.
// in webpack.config.js 
```
plugins: [
...
// new webpack.IgnorePlugin(/^\.\/locale$/, /moment$/),
// replace the line above with blow, regex to match the locale you want to use
new webpack.ContextReplacementPlugin(/moment[\/\\]locale/,  /(en|zh-cn)\.js/)
...
]
```

**Comment by JKarlavige** [comment]
It appears there's an issue with their latest version. I had to revert back to "2.24.0" without the "^" and that allowed my sites to build.

**Comment by drmats** [comment]
Same problem here with moment version `2.25.0` in one of my gatsby-based projects:
```
Cannot find module './locale'                              
                                                                                                                                              
                                                                                                                                              
> 1 | import { Moment } from './constructor';                                                                                                 
    | ^                                                                                                                                       
  2 |                                                                                                                                         
  3 | var proto = Moment.prototype;                                                                                                           
  4 |                                                                                                                                         
                                                                                                                                              
                                                                                                                                              
  WebpackError: Cannot find module './locale'                                                                                                 
                                                                                                                                              
  - prototype.js:1 webpackMissingModule                                                                                                       
    node_modules/moment/src/lib/moment/prototype.js:1:1                                                                                       
                                                                                                                                              
  - prototype.js:1 Module../node_modules/moment/src/lib/moment/prototype.js                                                                   
    node_modules/moment/src/lib/moment/prototype.js:1:1                                                                                       
                                                                                                                                              
  - moment.js:1 Module../node_modules/moment/src/lib/moment/moment.js                                                                         
    node_modules/moment/src/lib/moment/moment.js:1:1                                                                                          
                                                                                                                                              
  - moment.js:1 Module../node_modules/moment/src/moment.js                                                                                    
    node_modules/moment/src/moment.js:1:1
```

With moment `2.24.0` there is no such issue.

**Comment by ichernev** [comment]
Can someone provide a minimal (non)-working example so this can be investigated?

**Comment by paulohpf** [comment]
i've had to revert to "2.24.0", moment can't find './locale'

**Comment by qiqiboy** [comment]
Could you just unpublish 2.25.0?

**Comment by guigeekz** [comment]
this works perfectly for me : https://github.com/moment/moment/issues/4216#issuecomment-622453248

**Comment by NathanChannon** [comment]
> Can someone provide a minimal (non)-working example so this can be investigated?

Hey - please find attached. Just run npm install, and npm start :)

[issue5484.zip](https://github.com/moment/moment/files/4565276/issue5484.zip)

**Comment by ichernev** [comment]
Handling of require was reverted to 2.24.0 behavior.

Fixed in 2.25.1.

**Comment by bessarabov** [comment]
I've updated to the version `2.25.1`, but I still got the error. 

```
prototype.js:1 Uncaught Error: Cannot find module './locale'
    at webpackMissingModule (prototype.js:1)
    at Module../node_modules/moment/src/lib/moment/prototype.js (prototype.js:1)
    at __webpack_require__ (bootstrap:782)
    at fn (bootstrap:150)
    at Module../node_modules/moment/src/lib/moment/moment.js (moment.js:1)
    at __webpack_require__ (bootstrap:782)
    at fn (bootstrap:150)
    at Module../node_modules/moment/src/moment.js (moment.js:1)
    at __webpack_require__ (bootstrap:782)
    at fn (bootstrap:150)
    at Module../src/Menu.js (Home.js:186)
    at __webpack_require__ (bootstrap:782)
    at fn (bootstrap:150)
    at Module../src/App.js (Add.js:157)
    at __webpack_require__ (bootstrap:782)
    at fn (bootstrap:150)
    at Module../src/index.js (index.css?02e3:45)
    at __webpack_require__ (bootstrap:782)
    at fn (bootstrap:150)
    at Object.0 (serviceWorker.js:135)
    at __webpack_require__ (bootstrap:782)
    at checkDeferredModules (bootstrap:45)
    at Array.webpackJsonpCallback [as push] (bootstrap:32)
    at main.chunk.js:1
```

But maybe I'm doing something wrong. Can anybody confirm that this error has disappeared in 2.25.1 ?

**Comment by puopg** [comment]
Same issue for me and moment-timezone. Ive tried forcing 2.24.0, but still same issue since im guessing moment-timezone uses the latest?

**Comment by strange1120** [comment]
Still getting the error

**Comment by Fi1osof** [comment]
@strange1120 is moment@2.25.1?

**Comment by NathanChannon** [comment]
@ichernev Sorry, I can confirm 2.25.1 throws the same error. Using the app I uploaded previously, and a clean node_modules/package-lock.json, I can see the version is 2.25.1 but still throws.

**Comment by agrohs** [comment]
https://github.com/moment/moment/issues/5489

**Comment by agrohs** [comment]
Why are these issues closed...this is wreaking havoc

**Comment by CT83** [comment]
Yeah me too. I thought it was my Docker

**Comment by NathanChannon** [comment]
For anyone following this thread, I think for now, you can use [Shrinkwrapping](https://nodejs.org/en/blog/npm/managing-node-js-dependencies-with-shrinkwrap/) to block the latest dependency for whichever package may be using moment@latest

**Comment by nathanagez** [comment]
For people using create-react-app and who don't want to eject CRA you can install:

- [customize-cra](https://github.com/arackaf/customize-cra)
- [react-app-rewired](https://github.com/timarney/react-app-rewired)

Then create a `config-overrides.js` at root directory of your project for further overriding.
```js
module.exports = function override(config, env) {
  // do stuff with the webpack config...
  return config;
};
```
Edit your package.json:
```json
"start": "react-app-rewired start",
"build": "react-app-rewired build",
"test": "react-app-rewired test",
```
Because I use ant design my `config-overrides.js` looks like this:

```js
const { override, fixBabelImports } = require("customize-cra");
const webpack = require("webpack");

const fixMomentJs = (config) => {
  config.plugins.push(
    new webpack.ContextReplacementPlugin(/^\.\/locale$/, (context) => {
      if (!/\/moment\//.test(context.context)) {
        return;
      }
      // context needs to be modified in place
      Object.assign(context, {
        // include locales
        regExp: /^\.\/(fr|en)/,
        // point to the locale data folder relative to moment's src/lib/locale
        request: "../../locale",
      });
    })
  );
  return config;
};

module.exports = override(
  fixMomentJs,
  fixBabelImports("import", {
    libraryName: "antd",
    libraryDirectory: "es",
    style: "css",
  })
);
```
But yours will be more simple, something like this I think:
```js
module.exports = function override(config, env) {
  config.plugins.push(
    new webpack.ContextReplacementPlugin(/^\.\/locale$/, (context) => {
      if (!/\/moment\//.test(context.context)) {
        return;
      }
      // context needs to be modified in place
      Object.assign(context, {
        // include locales
        regExp: /^\.\/(fr|en)/,
        // point to the locale data folder relative to moment's src/lib/locale
        request: "../../locale",
      });
    })
  );
  return config;
};
```

**Comment by talgat-ruby** [comment]
Until the fix will be released, as was mentioned by @saadzr, you can temprorary just assign moment to "2.24.0" version. Remove carret: "^".

For those who does not know how to do it. Find the packages which uses moment: `npm ls moment`. You will probably see:
```
└─┬ antd@4.2.0
  ├── moment@2.25.1 
  └─┬ rc-picker@1.4.10
    └── moment@2.25.1  deduped
```

So inside `antd/package.json` and `rc-picker/package.json` in `dependencies`:
```
"moment": "^2.24.0", -> "moment": "2.24.0"
```

**Comment by nathanagez** [comment]
@talgat-ruby This implies to edit the node_modules folder that is not pushed on repositories, people who clone the project will have to do this workaround. I think this is a not a good workaround. 
For example we work with multiple collaborators and we can't told them to do this each time they clone a repository.
The best way to fix this until a fix is released is to configure Webpack.

**Comment by talgat-ruby** [comment]
@nathanagez this temprorary fix. Antd team will probably release official fix soon. And yes you are right you can not rely on it.

**Comment by nathanagez** [comment]
@talgat-ruby I explained how to configure Webpack, see upper, so you can push the workaround on your repository.

**Comment by manyu27** [comment]
This worked: 
https://github.com/moment/moment/issues/4216#issuecomment-622453248

**Comment by maxfeldman99** [comment]
i can confirm that @talgat-ruby suggestion fixed the problem for me.

**Comment by Jean-MarcCheynier** [comment]
@rilyu Your suggestion worked for me!

I am using moment-timezone and didn't specified a version for moment. 
It seems then, during the install, yearn installs the latest version of moment as a dependency -> moment@2.25 and it's not compatible.

Indeed adding 
"resolutions": { "moment": "2.24.0" }
solves the problem.

big thanks

**Comment by antonpod** [comment]
Any news from the devs?

**Comment by megphillips91** [comment]
Getting this error using MaterialUI DatePicker. Tried rolling back manually to 2.24.0 but no luck for me. Hope to see this resolved soon. I guess I will use the native TextField type="date" for now.

**Comment by GiovaniUrnau** [comment]
> Same issue,
> 
> Fixed by removing carret sign before version number. Package.json entry now looks like:
> "moment": "2.24.0"
> 
> It looks like a bug from the new version.

This works for me! Tnx

**Comment by faizhameed** [comment]
I am having the same problem i have deleted node_modules folder then changed the version of moment to 2.24.0 and then again I installed, the issue persists.
When I check rc-calendar is using 2.25.1 something like this. I tried editing there to and did `yarn install` still problem persists. Please help I'm working on an important project

**Comment by Jean-MarcCheynier** [comment]
@faizhameed Try to remove node-modules folder.
Remove your package-lock.json as well ( yarn.lock also if you're using yarn ).

Add this to your package.json after dependencies: {
 ...
},
resolutions: { "moment": "2.24.0" }

It Should force rc-datepicker to use moment 2.24

**Comment by pcarballeda** [comment]
Same problem here after upgrading to 2.25.0 and 2.25.1. 
The only solution that worked for me is downgrade to 2.24.0

**Comment by barrynorman** [comment]
Unfortunately I have the same problem. For some reason, the solution in the package.json is not working for me.
`resolutions: { "moment": "2.24.0" }`

For now I just downgraded like @pcarballeda

**Comment by anikkon** [comment]
For me fixing the version to `"moment": "2.24.0"` and adding `resolutions: { "moment": "2.24.0" }` temporarily solved the problem.

**Comment by Shakxzod-Namazbaev** [comment]
Module not found: Error: Can't resolve './locale' in 'D:\Downloads\Telegram\ildam\node_modules\rc-picker\node_modules\moment\src\lib\locale'

**Comment by Shakxzod-Namazbaev** [comment]
How can this be overcome?

  "moment": "2.24.0"

 "dependencies": {
    "antd": "^4.2.0",
    "axios": "^0.19.2",
    "classnames": "^2.2.6",
    "cross-env": "^7.0.2",
    "dva": "^2.4.1",
    "dva-model-extend": "^0.1.2",
    "less-vars-to-js": "^1.3.0",
    "libphonenumber-js": "^1.7.50",
    "lodash": "^4.17.15",
    "lodash.clonedeep": "^4.5.0",
    "moment": "2.24.0",
    "nprogress": "^0.2.0",
    "path-to-regexp": "^6.1.0",
    "prop-types": "^15.7.2",
    "query-string": "^6.12.1",
    "react": "^16.13.1",
    "react-currency-format": "^1.0.0",
    "react-custom-scrollbars": "^4.2.1",
    "react-dom": "^16.13.1",
    "react-helmet": "^6.0.0",
    "react-perfect-scrollbar": "^1.5.8",
    "svg-sprite-loader": "^4.2.1",
    "umi-plugin-locale": "^2.11.5"
  },

**Comment by MNoumanAbbasi** [comment]
Downgrading to 2.24.0 worked for me. Remove the ^ with 2.24.0.
Also manually updated all the moment dependencies to 2.24.0 in package-lock.json as well.

**Comment by pagoru** [comment]
Having the issue with the 2.24, trying to put the "resolutions"... but I didn't update to 2.25, what's wrong? It was working Thursday.

**Comment by Shakxzod-Namazbaev** [comment]
> Downgrading to 2.24.0 worked for me. Remove the ^ with 2.24.0.
> Also manually updated all the moment dependencies to 2.24.0 in package-lock.json as well.

Thanks worked! There is a problem with Yarn. Worked with Npm

**Comment by pagoru** [comment]
> Until the fix will be released, as was mentioned by @saadzr, you can temprorary just assign moment to "2.24.0" version. Remove carret: "^".
> 
> For those who does not know how to do it. Find the packages which use moment: `npm ls moment`. You will probably see:
> 
> ```
> └─┬ antd@4.2.0
>   ├── moment@2.25.1 
>   └─┬ rc-picker@1.4.10
>     └── moment@2.25.1  deduped
> ```
> 
> So inside `antd/package.json` and `rc-picker/package.json` in `dependencies`:
> 
> ```
> "moment": "^2.24.0", -> "moment": "2.24.0"
> ```
> 
> And run `npm i` TWICE. Do not know why need twice, but it only works after twice.

Working! <3

**Comment by purefan** [comment]
`yarn remove` + `yarn add moment` installed version 2.25.2 which builds fine for me

**Comment by Shakxzod-Namazbaev** [comment]
> `yarn remove` + `yarn add moment` installed version 2.25.2 which builds fine for me

Wouldn’t that be a problem in this version?

**Comment by purefan** [comment]
@Shakxzod-Namazbaev I think the original problem was with version 2.25.0, some people reported that the same problem happens in 2.25.1, others have mentioned that downgrading to 2.24 works for them. In my case version 2.25.2 seems to work, so maybe downgrading is not needed

**Comment by Shakxzod-Namazbaev** [comment]
> @Shakxzod-Namazbaev I think the original problem was with version 2.25.0, some people reported that the same problem happens in 2.25.1, others have mentioned that downgrading to 2.24 works for them. In my case version 2.25.2 seems to work, so maybe downgrading is not needed

So you don't have a problem with version 2.25.2 right now, right?

**Comment by purefan** [comment]
@Shakxzod-Namazbaev exactly, 2.25.2 works fine for me.

**Comment by drmats** [comment]
I can confirm - moment `2.25.2` works also for me.

**Comment by NathanChannon** [comment]
I think 2.25.2 is OK in standalone form but in packaged-managed form, it's still broken (webpack in my case where moment is a dependency of moment-timezone).

2.24.0 is still the one for people using webpack and other moment-dependant packages.

**Comment by drmats** [comment]
ver. `2.25.2` works for me in package-managed form - webpack configured in a gatsby-based project (ver. `2.25.0` didn't):

relevant `package.json` lines:
```
dependencies: {
    ...
    "moment": "^2.25.2",
    ...
}
```
I'm not using `moment-timezone` though.

**Comment by agrohs** [comment]
So crazy that this Issue report is still sitting as Closed with so many people affected and the only "solution" so far is to downgrade to an old version!?!?!

**Comment by bessarabov** [comment]
@ichernev Thank you! Updating to 2.25.2 fixes this issue for me!

```diff
--- a/front/package.json
+++ b/front/package.json
@@ -5,7 +5,7 @@
   "dependencies": {
-    "moment": "^2.22.2",
+    "moment": "^2.25.2",
```

**Comment by drmats** [comment]
OK. I praised too early. Although with moment ver. `2.25.2` webpack is able to produce working bundle without errors, but importing locale doesn't work anymore - looks like it's being ignored (perhaps this has something to do with [sideEffects / tree shaking](https://webpack.js.org/guides/tree-shaking/)?).

E.g.:

```javascript
import moment from "moment"
import "moment/locale/pl"

// set global 'moment' locale (polish)
moment.locale("pl")

// ... somewhere later
console.log(moment().format("DD MMMM YYYY HH:mm:ss"))
```

With version `2.25.2` output is **`04 May 2020 16:38:15`**, which is incorrect, as it should be
**`04 maja 2020 16:38:58`**.

So, version `2.24.0` is still the last working one.

**Comment by shubhamagarwal** [comment]
I faced the same issue, removing carrot from the version fixed this.

"moment": "2.24.0"

Version : 2.24.0

**Comment by ichernev** [comment]
Well, I was testing locally, and I can't really tell why, but if the is a package.json property `module` then it's broken. If there is only `jsnext:main` it is working. This relates to locale files not being loaded properly, while still included in the bundle.

So this "modern" module property that suppersedes jsnext:main is not just a rename, or an alt name. It has some hidden agenda that is not well advertised.

**Comment by ichernev** [comment]
2.25.3 is out without `module`.

**Comment by drmats** [comment]
@ichernev - thanks! I can confirm version `2.25.3` works (webpack is able to produce bundle without errors and including/setting locale works as expected).

**Comment by imatteh** [comment]
Fixes in 2.25.3

**Comment by Fhyrel** [comment]
> Same issue,
> 
> Fixed by removing carret sign before version number. Package.json entry now looks like:
> "moment": "2.24.0"
> 
> It looks like a bug from the new version.

After that works!
